### PR TITLE
[k8s] Bind pod ip to headless service

### LINKED ISSUE #3788: [k8s] multinode torch distributed nccl timeout
I tried running `examples/torch_ddp_benchmark` on kubernetes but the tasks hangs with the following error until throwing an NCCL timeout. It might be related to this [issue](https://github.com/pytorch/pytorch/issues/85300)?

```
(head, rank=0, pid=5752) 10.40.3.5
(head, rank=0, pid=5752) [W726 02:14:09.077658670 socket.cpp:697] [c10d] The IPv6 network addresses of (torch-bench-k8s-ebd1-3458-worker, 43913) cannot be retrieved (gai error: -2 - Name or service not known).
```

`task.yaml`
```
name: torch-ddp-bench

num_nodes: 2

resources:
  image_id: docker:nvcr.io/nvidia/pytorch:23.10-py3
  accelerators: H100:8 # Make sure you use 8 GPU instances
  cloud: kubernetes

file_mounts: 
  ./torch_ddp_benchmark.py: ./examples/torch_ddp_benchmark/torch_ddp_benchmark.py

setup: |
  conda activate ddp
  if [ $? -eq 0 ]; then
    echo 'conda env exists'
  else
    conda create -n ddp python=3.10 -y
    conda activate ddp
  fi
  pip install torch torchvision

run: |
  conda activate ddp
  num_nodes=`echo "$SKYPILOT_NODE_IPS" | wc -l`
  master_addr=`echo "$SKYPILOT_NODE_IPS" | head -n1`
  echo $master_addr
  torchrun \
    --nproc_per_node 8 \
    --rdzv_id=1 --rdzv_endpoint=${master_addr}:1234 \
    --rdzv_backend=c10d --nnodes $num_nodes \
    torch_ddp_benchmark.py --distributed-backend nccl
```

_Version & Commit info:_
* `sky -v`: `skypilot, version 1.0.0-dev0`
* `sky -c`: `skypilot, commit 93b319e8b90ab60d80f37e06802b876246a8e9fd`
* 1.29.6-gke.1326000

### PR DESCRIPTION
**Pull request by asaiacai**
Closes #3788 and #3510 

This PR:
- Renames worker pods so they are enumerated by hostname
- Creates a headless service per pod so pod IPs can be resolved by pod hostname via kube DNS.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
- [x] All smoke tests: `pytest tests/test_smoke.py --kubernetes --lf` 
- [x] `pytest tests/test_smoke.py::test_kubernetes_ssh_hostname --kubernetes`
- [x] `sky launch -c torch-ddp-bench --cloud kubernetes examples/torch-ddp-bench/torch-ddp-bench.yaml`
- [x] Manual checked created pod names and svc ips
```bash
$ sky local up
$ sky launch --num-nodes 3 --cpus 1 -c test -y 
$ kubectl get pod -o wide
$ kubect get svc
```

### CODE REVIEW

File: sky/provision/kubernetes/instance.py [hunk 1]
```diff
@@ -482,17 +481,19 @@ def _create_pods(region: str, cluster_name_on_cloud: str,
...
             head_selector = head_service_selector(cluster_name_on_cloud)
             pod_spec['metadata']['labels'].update(head_selector)
             pod_spec['metadata']['name'] = f'{cluster_name_on_cloud}-head'
         else:
             pod_spec['metadata']['labels'].update(constants.WORKER_NODE_TAGS)
-            pod_uuid = str(uuid.uuid4())[:4]
-            pod_name = f'{cluster_name_on_cloud}-{pod_uuid}'
-            pod_spec['metadata']['name'] = f'{pod_name}-worker'
+            pod_name = f'{cluster_name_on_cloud}-worker{pod_id}'
+            if pod_id == 0 or pod_name in running_pods:
```
Comment (romilbhardwaj) [inline sky/provision/kubernetes/instance.py:493]: good catch on handling this!

File: sky/provision/kubernetes/instance.py [hunk 2]
```diff
@@ -482,17 +481,19 @@ def _create_pods(region: str, cluster_name_on_cloud: str,
...
             pod_spec['metadata']['labels'].update(constants.HEAD_NODE_TAGS)
             head_selector = head_service_selector(cluster_name_on_cloud)
             pod_spec['metadata']['labels'].update(head_selector)
             pod_spec['metadata']['name'] = f'{cluster_name_on_cloud}-head'
         else:
             pod_spec['metadata']['labels'].update(constants.WORKER_NODE_TAGS)
-            pod_uuid = str(uuid.uuid4())[:4]
-            pod_name = f'{cluster_name_on_cloud}-{pod_uuid}'
-            pod_spec['metadata']['name'] = f'{pod_name}-worker'
+            pod_name = f'{cluster_name_on_cloud}-worker{pod_id}'
```
Comment (romilbhardwaj) [inline sky/provision/kubernetes/instance.py:492]: One of the reasons we used uuid is because the SSH config entries do not directly map to worker id. E.g., on this PR, I ran `sky launch -c test2 --cloud kubernetes --num-nodes 3 -- echo hi` followed by `ssh test2-worker1`, and I was ssh'd into test2-worker2 pod.

```
$ ssh test2-worker1
# Logged into pod named worker2
(base) sky@test2-2ea4-worker2:~$
```

Perhaps we need some handling in how SSH configs are generated?

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @asaiacai!

File: sky/provision/kubernetes/instance.py (hunk 2 above)
Comment (asaiacai) [reply sky/provision/kubernetes/instance.py:492]: ok so looks what's happening is when the ssh config is being generated, the ips are sorted [here](https://github.com/skypilot-org/skypilot/blob/1c32aa4ad349e4dc1f38e2e24537bc0b6ebfdae9/sky/backends/cloud_vm_ray_backend.py#L2338). So if two pods are created with the following mapping
```
worker1 10.44.3.19
worker2 10.44.1.20
```
the sort incorrectly flips the mapping which also explains why it's non-deterministic.
```
worker1 10.44.1.20
worker2 10.44.3.19
```

Question regarding `update_cluster_ips` in `cloud_vm_ray_backend.py` for `kubernetes`? Is it always going to use `internal_ips`? I imagine there's no external ips associated with pods. Just wondering if it's possible to just skip the sort step altogether as the logic seems tied to using external ips.

Comment (asaiacai) [reply sky/provision/kubernetes/instance.py:492]: fixed by removing the sorting logic in `update_cluster_ips`. resolving

**Comment by asaiacai** [comment]
@romilbhardwaj reran smoke tests and manual tests. Also added a new smoke test `test_kubernetes_ssh_hostname` to address the hostname mismatch bug. lmk if I can add anything else.

**Comment by zaptrem** [comment]
Thanks for the contribution @asaiacai! I tried it out but am still getting the ipv6 error

I 10-26 18:45:40 log_lib.py:415] Start streaming logs for job 1.
INFO: Tip: use Ctrl-C to exit log streaming (task will not be killed).
INFO: Waiting for task resources on 2 nodes. This will block if the cluster is full.
INFO: All task resources reserved.
INFO: Reserved IPs: ['10.0.40.102', '10.0.43.40']
(head, rank=0, pid=7879) Cluster name: amazon-eks-rmsnorm-dual-20-7a79
(head, rank=0, pid=7879) Master address: amazon-eks-rmsnorm-dual-20-7a79-head
(head, rank=0, pid=7879) Node rank: 0
(worker1, rank=1, pid=7055, ip=10.0.43.40) Cluster name: amazon-eks-rmsnorm-dual-20-7a79-worker1
(worker1, rank=1, pid=7055, ip=10.0.43.40) Master address: amazon-eks-rmsnorm-dual-20-7a79-worker1-head
(worker1, rank=1, pid=7055, ip=10.0.43.40) Node rank: 2
(worker1, rank=1, pid=7055, ip=10.0.43.40) [W1026 18:45:44.272302256 socket.cpp:697] [c10d] The IPv6 network addresses of (amazon-eks-rmsnorm-dual-20-7a79-worker1-head, 29400) cannot be retrieved (gai error: -2 - Name or service not known).

**Comment by asaiacai** [comment]
@zaptrem can you share your task yaml and the output of 

kubectl get svc

the PR should let you ping from one pod to another but if that’s not the case then I need to patch something.

**Comment by zaptrem** [comment]
> @zaptrem can you share your task yaml and the output of
> 
> kubectl get svc
> 
> the PR should let you ping from one pod to another but if that’s not the case then I need to patch something.

Actually I think it might have been some boneheaded ChatGPT IP-rewriting code I forgot to remove after switching to your branch. It appears they can see each other now, just only over TCP and not via EFA. I've been working on this for basically >24 hours straight (damn AWS and their 4:30am capacity block start times) so am not completely sure what the cause was. Thanks for the PR anyway!

Also, do you by any chance have a skypilot job definition/config that is working with Amazon's EFA you could share?

**Comment by asaiacai** [comment]
I’m actually in the same boat right now haha. I have a capacity reservation this week. Will update here if I figure it out

I also hate the 4:30am start times haha

**Comment by zaptrem** [comment]
Alright I figured it out so you don't have to stay up from 4am to 7pm doing so.

Build your skypilot image https://gist.github.com/zaptrem/de59394cc13b1ed298be89539c269906

(or use mine: https://hub.docker.com/repository/docker/zaptrem/nccl-tests/general but watch out for my secret backdoors)

```
EFA_INSTALLER_VERSION=1.35.0
AWS_OFI_NCCL_VERSION=v1.12.1-aws
NCCL_VERSION=v2.23.4-1
NCCL_TESTS_VERSION=v2.13.10

sudo docker build -f nccl-tests.Dockerfile \
       --build-arg="EFA_INSTALLER_VERSION=${EFA_INSTALLER_VERSION}" \
       --build-arg="AWS_OFI_NCCL_VERSION=${AWS_OFI_NCCL_VERSION}" \
       --build-arg="NCCL_VERSION=${NCCL_VERSION}" \
       --build-arg="NCCL_TESTS_VERSION=${NCCL_TESTS_VERSION}" \
       -t zaptrem/nccl-tests:v3 \
       .
```

Deploy wherever you want



Terraform your EKS cluster (you will need to give yourself 65 gazillion AWS IAM permissions)

main.tf
https://gist.github.com/zaptrem/c2fe513161f3d05180ff33eb4053a731

locals.tf
https://gist.github.com/zaptrem/9ec8091e6df5c00858ebf25816440b92

providers.tf
https://gist.github.com/zaptrem/558a3e6e091e5c4c96c0fef3ea6f542a

https://gist.github.com/zaptrem/dacf9d5e13c9d9f979d62ebe78853573




Update your local kubefig so it can see the cluster

`aws eks update-kubeconfig --name p5e-cluster-2 --region us-east-2`

Rename the context so SkyPilot doesn't freak out cuz there's a colon in the name
`kubectl config rename-context "arn:aws:eks:us-east-2:058264146096:cluster/p5e-cluster-2" "p5e-cluster-2"`


label your GPU nodes (my latest TF may automate this but maybe not idk, also if the tf starts crashing it's probably my gpu autolabelling code, didn't get to test it)

`kubectl label nodes ip-10-0-33-150.us-east-2.compute.internal skypilot.co/accelerator=h200 --overwrite`





add this to your skypilot config (or don't idk if it makes a difference)
```yaml

experimental:
  config_overrides:
    kubernetes:
      pod_config:
        spec:
          containers:
          - resources:
              requests:
                hugepages-2Mi: "5128Mi"
                vpc.amazonaws.com/efa: 32
              limits:
                hugepages-2Mi: "5128Mi"
                vpc.amazonaws.com/efa: 32
            securityContext:
              privileged: true
```



add this to your run script (I think this really matters)
```yaml
  export NCCL_DEBUG=INFO
  NCCL_SOCKET_IFNAME=eth0
  export FI_EFA_USE_HUGE_PAGE=0
```



Enjoy (or more likely cry as it doesn't work for you for some reason)

**Comment by romilbhardwaj** [comment]
Thanks for putting this together @zaptrem! This is awesome, would you like to put it together in a quick readme under something like `examples/efa`? :)

Some quick notes:

> Rename the context so SkyPilot doesn't freak out cuz there's a colon in the name

What version of SkyPilot are you on? We recently fixed handling for special characters in context name, so this step should not be necessary. I tested with the context name in your example. 

> label your GPU nodes

Curious, did you get a chance to try the [SkyPilot GPU labelling script](https://skypilot.readthedocs.io/en/latest/reference/kubernetes/kubernetes-setup.html#automatically-labelling-nodes)? Wondering if there's anything to be done to make it support H200s. 

> add this to your skypilot config

Note that adding `privileged: True` will break GPU isolation between pods if you plan on running more than one pod per node. Might be ok if you're running just one SkyPilot pod on each node.

**Comment by asaiacai** [comment]
@romilbhardwaj 

> What version of SkyPilot are you on? We recently fixed handling for special characters in context name, so this step should not be necessary. I tested with the context name in your example.

i think he's referring to my branch in this PR which was created before the cluster context name patch.

**Comment by zaptrem** [comment]
> Thanks for putting this together @zaptrem! This is awesome, would you like to put it together in a quick readme under something like `examples/efa`? :)
> 

Sure, ping me with this PR is merged (since I think it's needed for the pods to see each other).

> 
> What version of SkyPilot are you on? We recently fixed handling for special characters in context name, so this step should not be necessary. I tested with the context name in your example.

I'm on this branch which is a few months behind

> > label your GPU nodes
> 
> Curious, did you get a chance to try the [SkyPilot GPU labelling script](https://skypilot.readthedocs.io/en/latest/reference/kubernetes/kubernetes-setup.html#automatically-labelling-nodes)? Wondering if there's anything to be done to make it support H200s.

I did, and it got into a crash loop though I didn't investigate why. It's so easy/fast to do it myself locally I don't see the need for a pod for it. Maybe just run kubectl list + the label commands from local SkyPilot when doing sky check/etc?

> > add this to your skypilot config
> 
> Note that adding `privileged: True` will break GPU isolation between pods if you plan on running more than one pod per node. Might be ok if you're running just one SkyPilot pod on each node.

That is currently my setup since I'm doing distributed training.
